magento/devdocs#: Docs: Create topic for addDownloadableProductsToCart mutation (#5672)
magento/devdocs: guides/v2.3/graphql/mutations/add-downloadable-products.md

---
group: graphql
title: addDownloadableProductsToCart mutation
contributor_name: Atwix
contributor_link: https://www.atwix.com/
---

> A downloadable product can be anything that you can deliver as a file, such as an eBook, music, video, software application, or an update. To add a downloadable product to a cart, you must provide the cart ID, the SKU, and the quantity. Depends on downloadable product configuration you must specify IDs of downloadable product links. You can also optionally provide  customizable options.

Reviewer: ```suggestion
A downloadable product can be anything that you can deliver as a file, such as an eBook, music, video, software application, or an update. To add a downloadable product to a cart, you must provide the cart ID, the SKU, and the quantity. Depends on downloadable product configuration you must specify IDs of downloadable product links. You can also optionally provide customizable options.
```

Author: Nice catch :)

**fixed**

            "links": [
              {
                "title": "Solo Power Circuit",
                "price": 14
              }
            ],
            "samples": [
              {
                "title": "Trailer #1",
                "sample_url": "https://<M2_INSTANCE>/downloadable/download/sample/sample_id/10/"
              },
              {
                "title": "Trailer #2",
                "sample_url": "https://<M2_INSTANCE>/downloadable/download/sample/sample_id/11/"
              },
              {
                "title": "Trailer #3",
                "sample_url": "https://<M2_INSTANCE>/downloadable/download/sample/sample_id/12/"
              }
            ]
          }
        ]
      }
    }
  }
}
```

## Input attributes

The top-level `AddDownloadableProductsToCartInput` object is listed first. All child objects are listed in alphabetical order.

### AddDownloadableProductsToCartInput object {#AddDownloadableProductsToCartInput}

The `AddDownloadableProductsToCartInput` object must contain the following attributes:

Attribute |  Data Type | Description
--- | --- | ---
`cart_id` | String! | The unique ID that identifies the customer's cart

> `cart_items` | [DownloadableProductCartItemInput!](#DownloadableProductCartItemInput) | Contains the cart item IDs and quantity of each item

Reviewer: ```suggestion
`cart_items` | [[DownloadableProductCartItemInput!]!](#DownloadableProductCartItemInput) | Contains the cart item IDs and quantity of each item
```

Author: **fixed**

### DownloadableProductCartItemInput object {#DownloadableProductCartItemInput}

> The `DownloadableProductCartItemInput` object must contain the following attributes:

Reviewer: We say that the object must contain the following attributes, however, `downloadable_product_links` and `customizable_options` are optional attributes. So "must" is not the 100% correct assumption here.

Author: Really a good point.

**fixed**

Attribute |  Data Type | Description
--- | --- | ---

> `customizable_options` |[[CustomizableOptionInput]](#CustomizableOptionInputVirtual) | An array that defines customizable options for the product

Reviewer: ```suggestion
`customizable_options` |[[CustomizableOptionInput!]](#CustomizableOptionInputVirtual) | An array that defines customizable options for the product
```

Author: **fixed**

---
group: graphql
title: addDownloadableProductsToCart mutation
contributor_name: Atwix
contributor_link: https://www.atwix.com/
---

A downloadable product can be anything that you can deliver as a file, such as an eBook, music, video, software application, or an update. To add a downloadable product to a cart, you must provide the cart ID, the SKU, and the quantity. Depends on downloadable product configuration you must specify IDs of downloadable product links. You can also optionally provide customizable options.

## Syntax

> `mutation: {addDownloadableProductsToCart(input: AddDownloadableProductsToCartInput): {AddDownloadableProductsToCartOutput}}`

Reviewer 2: Suggest using multi-line `graphql` representation.

Author: **implemented**

## Example usage

These examples show how to add a downloadable product into shopping cart with enabled/disabled `Links can be purchased separately` downloadable product option.

### Add a downloadable product to a cart with enabled `Links can be purchased separately`

The following example displays how to add a downloadable product with `sku` value of `240-LV09` and with the custom downloadable links: `Episode 2`, `Episode 3` from the Luma sample data into shopping cart. The product with `sku` `240-LV09` has enabled `Links can be purchased separately` option, so you must specify downloadable links which will be added into cart.

**Request**

> ```text

Reviewer 2: ```suggestion
```graphql
```

Author: **implemented**

---
group: graphql
title: addDownloadableProductsToCart mutation
contributor_name: Atwix
contributor_link: https://www.atwix.com/
---

> A downloadable product can be anything that you can deliver as a file, such as an eBook, music, video, software application, or an update. To add a downloadable product to a cart, you must provide the cart ID, the SKU, and the quantity. Depends on downloadable product configuration you must specify IDs of downloadable product links. You can also optionally provide customizable options.

Author: **implemented**

## Syntax

```graphql
mutation: {
    addDownloadableProductsToCart(
        input: AddDownloadableProductsToCartInput
    ): {
        AddDownloadableProductsToCartOutput
    }
}
```

## Example usage

> These examples show how to add a downloadable product into shopping cart with enabled/disabled `Links can be purchased separately` downloadable product option.

Author: **implemented**

> ### Add a downloadable product to a cart with enabled `Links can be purchased separately`

Author: **implemented**

> The following example displays how to add a downloadable product with `sku` value of `240-LV09` and with the custom downloadable links: `Episode 2`, `Episode 3` from the Luma sample data into shopping cart. The product with `sku` `240-LV09` has enabled `Links can be purchased separately` option, so you must specify downloadable links which will be added into cart.

Author: **implemented**

          }
        }
      }
    }
  }
}
```

**Response**

```text
{
  "data": {
    "addDownloadableProductsToCart": {
      "cart": {
        "items": [
          {
            "product": {
              "sku": "240-LV09"
            },
            "quantity": 1,
            "links": [
              {
                "title": "Episode 2",
                "price": 9
              },
              {
                "title": "Episode 3",
                "price": 9
              }
            ],
            "samples": [
              {
                "title": "Trailer #1",
                "sample_url": "https://<M2_INSTANCE>/downloadable/download/sample/sample_id/16/"
              },
              {
                "title": "Trailer #2",
                "sample_url": "https://<M2_INSTANCE>/downloadable/download/sample/sample_id/17/"
              },
              {
                "title": "Trailer #3",
                "sample_url": "https://<M2_INSTANCE>/downloadable/download/sample/sample_id/18/"
              }
            ]
          }
        ]
      }
    }
  }
}
```

### Add a downloadable product to a cart with disabled `Links can be purchased separately`

> The following example displays how to add a downloadable product with the `sku` value of `240-LV07` from the Luma sample data into shopping cart. This product has disabled `Links can be purchased separately` option, so all downloadble links will be added into cart automatically. 

Reviewer 3: ```suggestion
The following example shows how to add a downloadable product in which the **Links can be purchased separately** option is disabled. All downloadable links are added to the cart automatically.
```

Author: **implemented**

            ],
            "samples": [
              {
                "title": "Trailer #1",
                "sample_url": "https://<M2_INSTANCE>/downloadable/download/sample/sample_id/10/"
              },
              {
                "title": "Trailer #2",
                "sample_url": "https://<M2_INSTANCE>/downloadable/download/sample/sample_id/11/"
              },
              {
                "title": "Trailer #3",
                "sample_url": "https://<M2_INSTANCE>/downloadable/download/sample/sample_id/12/"
              }
            ]
          }
        ]
      }
    }
  }
}
```

## Input attributes

The top-level `AddDownloadableProductsToCartInput` object is listed first. All child objects are listed in alphabetical order.

### AddDownloadableProductsToCartInput object {#AddDownloadableProductsToCartInput}

The `AddDownloadableProductsToCartInput` object must contain the following attributes:

Attribute |  Data Type | Description
--- | --- | ---
`cart_id` | String! | The unique ID that identifies the customer's cart
`cart_items` | [[DownloadableProductCartItemInput!]!](#DownloadableProductCartItemInput) | Contains the cart item IDs and quantity of each item

> ### DownloadableProductCartItemInput object {#DownloadableProductCartItemInput}

Reviewer 3: This section is not listed alphabetically.

Author: **fixed**

The `CartItemInput` object must contain the following attributes:

{% include graphql/cart-item-input.md %}

### CustomizableOptionInput object {#CustomizableOptionInputVirtual}

The `CustomizableOptionInput` object must contain the following attributes:

{% include graphql/customizable-option-input.md %}

### DownloadableProductCartItemInput object {#DownloadableProductCartItemInput}

> The `DownloadableProductCartItemInput` object must contain the following attribute:

Author: **fixed**

Attribute |  Data Type | Description
--- | --- | ---
`data` | [CartItemInput!](#CartItemInputVirtual) | An object containing the `sku` and `quantity` of the product

and the next optional ones:

> Attribute |  Data Type | Description

Author: **done**

--- | --- | ---
`customizable_options` |[[CustomizableOptionInput!]](#CustomizableOptionInputVirtual) | An array that defines customizable options for the product
`downloadable_product_links` | [[DownloadableProductLinksInput!]](#DownloadableProductLinksInput) | An object containing the `link_id` of the downloadable product link

### DownloadableProductLinksInput object {#DownloadableProductLinksInput}

> 

Author: **added**

Attribute |  Data Type | Description
--- | --- | ---
`link_id` | Int! | A unique ID (`downloadable_link`.`link_id`) of the downloadable product link

## Output attributes

The `AddDownloadableProductsToCartOutput` object contains the `Cart` object.

Attribute |  Data Type | Description
--- | --- | ---
`cart` |[Cart!](#CartObject) | Describes the contents of the specified shopping cart

### Cart object {#CartObject}

{% include graphql/cart-object.md %}

## Errors

Error | Description
--- | ---

> `Required parameter "cart_id" is missing` | Mutation `cart_id` parameter is empty

Reviewer 3: ```suggestion
`Required parameter "cart_id" is missing` | The mutation does not contain a `cart_id` parameter.
```

Author: **implemented**

> `Required parameter "cart_items" is missing` | Mutation `cart_items` parameter is empty or is not a type of `array`

Reviewer 3: ```suggestion
`Required parameter "cart_items" is missing` | The `cart_items` parameter is empty or is not of type `array`.
```

Author: **implemented**

> `Please specify product link(s).` | Error appears when you try to add a downloadable product with enabled `Links can be purchased separately` option and you did not specify links which you want to buy in the `mutation` query.

Reviewer 3: ```suggestion
`Please specify product link(s).` | You tried to add a downloadable product in which the `Links can be purchased separately` option is enabled, but you did not specify individual product links.
```

Author: **implemented**
